### PR TITLE
Added Subscription Cancellation Confirmation Notification Feature

### PR DESCRIPTION
**Pull request by Ayobamidele**
### Description
<hr>
Implement a feature to send Subscription Cancellation Confirmation to users. This endpoint should immediately update users about the Subscription Cancellation Confirmation details in the system. Upon being successfully sent, it will return the message confirming it is being delivered with a '200' status. If an error occurs, an appropriate error status will be returned.

### Acceptance Criteria
<hr>

- Subscription must already be canceled

 
### Purpose
<hr>
Users get a record/notification of their subscription cancellation confirmation

### Requirements
<hr>

- [x] Subscription exists and has been canceled.
- [x] Ensure only authorized users have access to the endpoint.

## Endpoints
     [POST] /api/v1/user/subscription/notification/cancellation/{id}

**Requests**
```
POST /api/v1/user/subscription/notification/cancellation/{id}
Content-Type: application/json
Authorization: Bearer <access_token>

{
    "id": 1
}

```

**Successful Response**

- Status code: 200

```json
     {
           "message": "Sent Subscription Cancellation Confirmation notification successfully",
           "name": "str",
           "subscription": {
               "id": "int",
               "plan": "str",
               "cancellation_date": "str"
           }
       }
```

**Error Response**

401 **Unauthorized**
```
{
  "status":  "Unauthorized",
  "message":  "Unauthorized. Please log in."
  “status-code”: 401
}
```

400 **Bad Request**
```
{
  "status":  "Bad Request",
  "message":  "Please check the submitted data"
  “status-code”: 400
}
```
404 **No subscription matches id.**
```
{
  "success": false,
  "message": "No Subscription Found",
  "status_code": 404
}
```
500 **Internal Server Error**

```
{
  "success": false,
  "message": "string",
  "status_code": 500
}
```

### Database design

schema

Table name: **Subscription**

id:

    constraints: string(uuid), unique, primary-key, not null

user_id:

    constraints: integer, not null

plan:

    constraints: string, not null, length(150)

is_active:
    
     constraint: boolean

cancelled_at:

    constraints: string, date timestamp

createdAt:

    constraints: string, date timestamp
     

**Tasks**:
 

- [x] Create the endpoint: `[POST] /api/v1/user/subscription/notification/cancellation/{id}`
- [x] Implement the server-side logic to handle subscription cancellations. This includes sending the confirmation notification.
- [x]  Set up logging to capture errors and notification attempts.
- [x]  Write unit tests to verify that the cancellation confirmation is sent correctly.

**Testing**:
   - [x] unit tests, cover scenarios such as valid subscription IDs, invalid subscription IDs, and edge cases.
   - [x] For security testing, consider aspects like input validation and authorized users.

Screenshot 2024-07-21 230208

### CODE REVIEW

File: api/v1/routes/subscription.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                    content=jsonable_encoder({
+                    "message": "Sent Subscription Cancellation Confirmation notification successfully",
+                    "name": subscription_user.username,
+                    "subscription": {
+                        "id": subscription.id,
+                        "plan": subscription.plan,
+                        "cancellation_date": subscription.end_date
+                    }
+                }),) 
+
```
Comment (Dev-wonderful) [inline api/v1/routes/subscription.py:31]: There's `JsonResponseDict` class in the `utils.json_response` module that handles this for you according to the specified convention, check it out and update all round

File: tests/v1/test_subsciption.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+TestingSessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
+
+def override_get_db():
+	db = TestingSessionLocal()
+	try:
+		yield db
+	finally:
+		db.close()
+
+app.dependency_overrides[get_db] = override_get_db
```
Comment (Dev-wonderful) [inline tests/v1/test_subsciption.py:25]: Find out how to mock database

File: api/v1/routes/subscription.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                    "status_code": 500
+                }),)
+
+
+
+@subscription.post("/notification/cancellation/{subscription_id}")
+def send_subscription_cancellation_confirmation(subscription_id: str, current_user:  User = Depends(get_current_user), db: Session = Depends(get_db)):
+    try:
+        invalid_request = SubscriptionRequest(id=subscription_id)
+        subscription = db.query(Subscription).filter_by(id=subscription_id).first()
```
Comment (Dev-wonderful) [inline api/v1/routes/subscription.py:48]: You should be able to put `uuid.UUID` as the type for `subscription_id` and FastApi would handle the validation for you

File: api/v1/models/subscription.py [hunk 4]
```diff
@@ -25,6 +25,7 @@ class Subscription(BaseModel, Base):
     user_id = Column(UUID(as_uuid=True), ForeignKey('users.id'), nullable=False)
     plan = Column(String(50), nullable=False)
     is_active = Column(Boolean, default=True)
+    plan_id = Column(Integer)
     start_date = Column(DateTime(timezone=True), server_default=func.now())
```
Comment (Dev-wonderful) [inline api/v1/models/subscription.py:29]: Is there a migration for this new field `plan_id`

**Review by Dev-wonderful — CHANGES_REQUESTED** [review]

File: api/v1/routes/subscription.py (hunk 1 above)
Comment (Ayobamidele) [reply api/v1/routes/subscription.py]: I've updated it

File: tests/v1/test_subsciption.py (hunk 2 above)
Comment (Ayobamidele) [reply tests/v1/test_subsciption.py]: Also updated

File: api/v1/routes/subscription.py (hunk 3 above)
Comment (Ayobamidele) [reply api/v1/routes/subscription.py]: I've corrected that

File: api/v1/models/subscription.py (hunk 4 above)
Comment (Ayobamidele) [reply api/v1/models/subscription.py:29]: yes there is

**Comment by Ayobamidele** [comment]
@Dev-wonderful  I've solved all issues
